### DISCUSSION TITLE
posts/656

### DISCUSSION THREAD

**Discussion by giscus** [kakaotech/comments#21] [Announcements]
# posts/656

안녕하세요, 데이터분석플랫폼 조직의 루이스입니다.
시리즈의 첫 글인 아파치 플링크...

https://tech.kakao.com/posts/656

**Comment by joswlv** [comment]
안녕하세요, 좋은 글 잘 읽었습니다! 글을 보다가 궁금한 점이 있어 이렇게 댓글을 남깁니다.

글에서 설명하신 쓰기 로직을 보면, upsert 모드가 활성화된 상태에서 INSERT가 많은 테이블의 경우 Iceberg sink 결과로 equality delete 파일이 많이 생성될 수 있고, 이로 인해 읽기 성능이 저하될 가능성이 있어 보입니다.
op 코드로만 구분해서 sink를 처리하고, 복구 시 리플레이를 위한 별도 모드를 추가해 upsert를 최소화하는 방식이 성능 측면에서 더 유리하지 않을까 생각했는데요. 

혹시 제가 놓친 부분이 있거나, 카카오만의 특수한 상황이 있어 이런 방식으로 구현하신 것이라면 공유해 주시면 글을 이해하는 데 큰 도움이 될 것 같습니다.

다시 한 번 좋은 글 작성해 주셔서 감사합니다!

**Reply by SML0127** [reply]
안녕하세요, joswlv~ 
루이스 이승민입니다.

먼저 긴 글 읽어주셔서 감사합니다.
글을 쓰다보니 공유하고 싶은 내용이 많아져 글이 길어졌는데, 끝까지 읽어 주시는 분들이 계셔서 너무 감사하고 행복합니다.

이제 질문 주신 내용들에 대해 제가 이해한 바를 기준으로 답변 드려봅니다.
> 글에서 설명하신 쓰기 로직을 보면, upsert 모드가 활성화된 상태에서 INSERT가 많은 테이블의 경우 Iceberg sink 결과로 equality delete 파일이 많이 생성될 수 있고, 이로 인해 읽기 성능이 저하될 가능성이 있어 보입니다. op 코드로만 구분해서 sink를 처리하고, 복구 시 리플레이를 위한 별도 모드를 추가해 upsert를 최소화하는 방식이 성능 측면에서 더 유리하지 않을까 생각했는데요. 혹시 제가 놓친 부분이 있거나, 카카오만의 특수한 상황이 있어 이런 방식으로 구현하신 것이라면 공유해 주시면 글을 이해하는 데 큰 도움이 될 것 같습니다.


먼저 카카오만의 특수한 상황이기에 sink 로직이 본 글과 같이 구현된건지 부분에 답변 드리겠습니다.
- sink 로직은 저희가 구현한게 아닌 apache iceberg 라이브러리의 Flink 적재 로직인 [basedeltataskwriter.java](https://github.com/apache/iceberg/blob/2b55fef7cc2a249d864ac26d85a4923313d96a59/flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/BaseDeltaTaskWriter.java#L79-L110)의 일부 입니다. 
- 라이브러리를 그대로 사용하기에, 카카오만의 특수한 상황은 아니에요.

다음으로는 제안주신 `복구 시 리플레이를 위한 별도 모드 추가 및 이를 통한 읽기 성능 저하 최소화`가 가능할지, 제 생각을 공유드리겠습니다. 
- INSERT 쿼리를 equality delete 파일에 넣는건 복구나 재처리등의 리플레이에 정말 유용하다 생각되지만, 말씀 주신 것처럼 읽기 성능이 느려 지는 필연이 생깁니다. 일종의 트레이드 오프라고 생각해요
- 또한 이슈나 문제라는게 언제 발생할지 모르기에, 실 환경에서 두 모드를 구분 해서 사용하기란 쉽지 않을 것 같습니다..

다만, 아래와 같은 방향이면 적재가 느려지고 메모리도 어느정도 점유하겠으나, upsert를 최소화하여 빠른 조회가 가능해보여요
- 동등 칼럼이 0 이상의 정수형 단일 칼럼으로 구성된 경우 한정합니다.
	- bitmap으로 들어온 동등 칼럼의 값을 유지(들어온 값을 인덱스로하여 해당 bit를 1로 set)
	- bitmap을 보고 기존에 들어온 값이라면 equality delete에 해당 내용을 추가(현 upsert와 동일하게 동작), 없다면 데이터 파일에만 적재
	- delete 쿼리가 들어온다면, bitmap에서 해당 값을 인덱스로한 bit를 0으로 unset
	  
- 10억개 레코드 기준, 110MB(1,000,000,000 / 8 / 1024 / 1024) 내외 크기이기에 메모리 유지도 가능해 보입니다.
	- 플링크 잡이 재시작되거나 죽었다 다시 시작하는 상황을 고려하여 플링크의 체크포인트에서 해당 정보를 유지
	  
- bitmap 대신 mysql의 인덱스처럼 b / b+ tree 구조도 생각해보았는데, 동등 칼럼 값이 일정하지 않고 띄엄띄엄 들어오는 경우에 더 적합할 것 같아요.

언제든지 편하게 답변 주셔도 되며, 긴 글 읽어주셔서 다시 한번 감사드려요.
좋은 하루 보내세요!

**Reply by SML0127** [reply]
+) 인덱스로 LSM 트리를 차용하여, 동등 필드 칼럼 값 유무를 확인하는 블룸 필터로 사용해도 좋을 것 같습니다. 다만.. 이러한 인덱스를 활용하는 방식이 실시간으로 데이터를 적재하는 상황에 적절한지는 의문이긴합니다  🤔

**Comment by 911432** [comment]


**Reply by SML0127** [reply]
@911432 
아래의 링크에서 수정 또는 삭제가 가능합니다!
https://github.com/kakaotech/comments/discussions/21

**Comment by 911432** [comment]
열심히 연구해주셔서 감사합니다.

Iceberg는 넷플릭스에서 제공하는 동영상 같이 파일의 변경이 작은 환경에서 적합합니다.
알리바바 클라우드에서 발표한 [내용](https://www.alibabacloud.com/blog/data-lake-for-stream-computing-the-evolution-of-apache-paimon_601270)을 참고가 될 것 같아요.
여기에 따르면 Iceberg보다 Paimon이 더 좋다고 합니다.

데이터베이스에 있는 로그를 FlinkCDC로 가지고 오고, Flink CDC에서 [FileSystem인 Catalog](https://paimon.apache.org/docs/master/concepts/catalog/#filesystem-catalog)와 [Paimon이라는 테이블 포멧을 가진 데이터](https://paimon.apache.org/)를 [카카오 클라우드에 있는 Object Storage 같은](https://docs.kakaocloud.com/service/bss/object-storage)  [S3 Complaint Object Stores](https://paimon.apache.org/docs/master/maintenance/filesystems/#s3-complaint-object-stores)에 저장하고, 이 S3에 저장된 Catalog와 데이터를 데이터를 원격으로 작업할 수 있게 만든 [Flink 2.x](https://flink.apache.org/2024/10/23/preview-release-of-apache-flink-2.0/#disaggregated-state-storage-and-management)인 컨테이너 이미지로 컨테이너 런타임을 통해 노드인 머신로 배포된 곳에 가지고 와서 분석할 수 있는 환경을 만들면 될 것 같습니다.

**Reply by SML0127** [reply]
@911432
공유해주신 링크들과 내용들이 
이후 방향성을 설립하는데 많은 참고가 되었습니다.

특히 paimaon 도입은 초창기 hudi와의 성능 비교 글을보고 회의적인 생각이 많았는데, 최근 내용들을 보면 더 좋은 선택지가 될 수 있다 생각이듭니다. 
정말 감사합니다!

**Comment by ml-taehoon-choi** [comment]
안녕하세요, 좋은글 감사합니다 :) 
위 대댓글보니 paimon과 hudi도 검토해보신 것 같고 paimon은 성능비교 글을 보고 회의적인 생각이 드신 것 같습니다.
그럼 hudi는 어떤 점 때문에 선택하지 않으신건가요?

**Reply by SML0127** [reply]
@ml-taehoon-choi 
안녕하세요~
먼저 긴글 읽어주셔서 감사합니다.😄😄

hudi를 선택하지 않은 이유는 사내에서 제공되는 플랫폼들과 긴밀하게 협업하기 위함입니다. 올 하반기 사내 trino에 iceberg 커넥터가 추가되어, trino를 통한 Iceberg 테이블 조회나 관련 지표 추출이 가능해 졌기 때문입니다. 해당 이유로 hudi는 고려하지 않았습니다.
("hudi와 iceberg 성능은 비슷한 수준" -> "paimon은 hudi보다 성능이 많이 안좋네" -> "사내에서는 Iceberg 지원으로 노선을 잡았네" -> "Iceberg 써야겠다" 순으로 생각했다고 봐주시면 될것같습니다🤣🤣)


tmi로 몇가지를 더 공유드려보면, 내년에는 data lakehouse 관련하여 두가지를 더 고려 중입니다.
- paimon의 경우 내년에 POC를 진행 해볼 예정입니다. paimon이 최근 iceberg 만큼 배치(Spark) 성능이 올라왔다고 들었으며, 스트리밍은 기존에도 매우 좋은 성능을 보인다하여 초당 수십만개의 로그가 들어오는 경우들에 한하여 좋은 선택지가 될 수 있을지 확인해보려합니다.
- iceberg table version v3가 2025년 준비 중인데 더 좋은 스트리밍 / CDC 기능을 지원하려한다고합니다. 이에 v3가 나오면 한번 테스트 해보려합니다.
<img width="812" alt="image" src="https://github.com/user-attachments/assets/1899487b-d3d1-4c12-8b75-881d217cb96a">

<br>
<br>


외에도 최근 해외에서는 data lakehouse 기술의 추세(?)가 Iceberg로 완전히 넘어 간것으로 보이는데, 관련하여 최근에 본 글 링크 [Apache Iceberg Won the Future — What’s Next for 2025?](https://blog.det.life/apache-iceberg-won-the-future-whats-next-for-2025-731635bfcb7a) 공유드리니, 시간되시면 한번 읽어 보시길 추천드립니다.(위에 이미지도 해당 글에서 캡처했어요)

추가로 궁금하신 내용 있으시면 편하게 남겨주시면 확인 후 답변드리겠습니다.
다시 한번 긴글 봐주셔서 정말 감사합니다.🙇🏻‍♂️🙇🏻‍♂️

**Comment by Haebuk** [comment]
안녕하세요~ 너무 좋은 글 잘 읽었습니다.
이렇게 세세히 공유해주시는 분이 계셔 후발대(?)에게 큰 도움이 됩니다.
감사합니다.

**Reply by SML0127** [reply]
감사합니다!!😃😃😃

**Comment by JeonDaehong** [comment]
안녕하세요! 좋은 글 너무 잘 읽었습니다!
저희도 아이스버그를 도입하려고 시도하고 있는데, 정말 많은 도움이 되었습니다.

혹시 질문을 하나 드려도 괜찮을까요!?ㅎㅎ

Compaction 작업을 Auto 로 수행하고자 하는데, 파일의 min, max 크기등으로 제어하는 것이 아니라
특정 시간으로 Compaction 을 자동으로 수행하게끔 하는 옵션은 없는지 궁금합니다!

감사합니다.

**Reply by SML0127** [reply]
안녕하세요, @JeonDaehong 

먼저 긴글 읽어주셔서 너무 감사합니다.
질문 주신 내용에대해 테스트해본바와 생각해본 내용 간략히 공유드려봅니다. 아이스버그 도입에 도움이 되었으면해욥!

- 본 글은 MySQL 테이블을 CDC 하여 UPSERT 모드로 Iceberg로 적재하며, 모든 파일(`rewrite-all`)에 대해 compaction을 수행하고 있습니다. DB 데이터를 UPSERT 모드로 적재하기에 과거 데이터가 수정, 삭제 될 수 있어 예외없이 모든 파일에 대해 수행해야 small file을 최소화할 수 있음을 다시 한번 공유드립니다.
- compaction 자동 수행은 보통 스케쥴러(e.g. airflow)와 처리 엔진(e.g. spark)의 조합으로 많이 사용합니다. 이외에 메인터넌스를 자동으로 제공하는 툴로는 apache amoro가 있는데, 제가 알기로 amoro는 파일 크기 조건 밖에 없는것으로 알아요.
- 특정 시간 기준 compaction이 두가지 관점이 있을 것 같습니다. 첫번째는 파일의 생성 / 수정 시간을 사용하는 방법, 두번째는 시간 관련 칼럼(e.g. created_at)을 사용하는 방법이 있을 것 같습니다.
- 아쉽게도 첫번째 방법은 현시점 존재하는 처리 엔진(e.g. spark)에서는 지원하지 않는 것으로 알고있고, 제공되는 Iceberg Java API 을 통해 직접 만드는 방법밖에 없어보입니다. 다만 이 방식의 경우 UPSERT로 데이터를 적재하면 과거 데이터 파일에서 참조되는 삭제 파일은 계속 참조 상태가 유지되기에, compaction 후보군에서 계속 제외되어 small file로써 계속 유지되는 이슈가 발생 가능합니다.🥲
- 두번째 방법은 저희가 서버 로그를 APPEND 모드로 Iceberg에 적재하는 것을 테스트할때 사용해보았습니다. 처리 시간 칼럼에 파티션(e.g. hour transform, identity transform)을 걸고, compaction 조건에 filter를 두어 원하는 시간대만 compaction을 수행하게 하였습니다. 예를 들어 14시 ~ 15시에 적재된 파티션 (e.g. `/data/created_at=2025-03-10+14`)을 타겟하도록 filter 조건을 두어 compaction을 수행한다면, 특정 시간으로 compaction 수행이 가능합니다. 저희 경우는 매 시간 10분마다 이전 시간대의 파티션에 메인터넌스를 수행하도록 설정하였습니다. 아래는 공식 문서의 필터 예시이니 참고하시면 좋을듯합니다.
<img width="777" alt="image" src="https://github.com/user-attachments/assets/dffe4c4a-8b21-49d8-b89e-0d91b50ad693" />


좋은 하루되세요!

**Reply by JeonDaehong** [reply]
답변해주셔서 감사합니다 !!

올려주신 글 덕분에, 아이스버그에 대해 많이 공부가 되었고, 아이스버그 도입에도 많은 도움이 되었습니다.
평소에도 관심이 많아서 kakao tech 잘 보고 있지만, 앞으로도 꾸준히 정독하겠습니다!

혹시.. 또 궁금한 점이 생기면 답글로 가끔 질문드려도 괜찮을까요..!?ㅎㅎㅎ

좋은 하루 되시기 바랍니다!

**Reply by SML0127** [reply]
제가 감사합니다!
궁금하신 사항 생기시면 언제든 편하게 질문 주셔요~

**Reply by permitroot** [reply]
안녕하세요~! 
우선 작성해주신 글을 읽고 도움이 많이 되었습니다. 
감사합니다 🙇🏻

글을 읽다가 궁금한 점이 있어 질문을 이어서 달게 되었습니다. 
위에 샤딩으로 같은 테이블에 여러 파티션에 upsert 작업을 하였을 때 conflict 이슈를 겪으신 부분을 확인했습니다. 
cdc 로 발생하는 upsert 와 주기적으로 별도로 실행되는 compaction maintain 과는 이런 충돌을 안겪으셨는지 궁금합니다. 관련해서 이슈 해결 경험이나 의견을 주신다면 너무 큰 도움이 될 것 같습니다 :)

**Reply by SML0127** [reply]
@permitroot 
안녕하세요!
먼저, 글을 읽어 주셔서 감사합니다.

질문 주신 사항에 답변드리겠습니다.
- commit 충돌은 간혹 발생하지만, 현재까지 이로 인해 iceberg 테이블에 데이터를 적재하는 flink job에 영향이 발생한 적은 없습니다.
- 잘 아실겠지만, Iceberg는 OCC(optimistic concurrency control)을 사용합니다. 즉, commit 직전 충돌 여부를 판단하고, 충돌이 발생하면 commit을 재시도하거나 유저가 정의한대로(e.g. 바로 실패) 동작합니다. 
- commit 충돌은 주로 하나의 테이블에 여러 사용자 동시에 작업을 시도할때 발생합니다. 말씀하신 것처럼, compaction(spark의 rewrite_data_file)도 commit을 수행하기에 충돌이 발생할 가능성이 있습니다. 특히 `partial-progress.enabled`를 true로 설정하면, 파일 그룹(= partition)에 대해 compaction이 완료 될때마다 commit이 수행되므로 충돌 가능성이 더 높아집니다.
- 이를 위해 저희는 commit 재시도 횟수를 기본값 4회에서 60회로 늘려 commit에 대한 내성은 올리되, commit의 total timeout을 30분에서 5분으로 낮추어 SLA를 만족시키는 방향으로 운영하고 있습니다.

위 답변이 도움이 되셨기를 바랍니다. 감사합니다!

**Reply by permitroot** [reply]
답변 너무너무 감사드립니다 ! 

행복한 하루 보내시길 바랍니다.

**Reply by JeonDaehong** [reply]
@SML0127

안녕하세요! 다시 아이스버그를 사용하던 중 궁금한점이 생겨 질문드립니다...!!

통계정보를 저장하는 퍼핀파일이라는게 있다고 하는데, 스파크 카탈로그를 사용할 때는 안만들어지는 거 같습니다.

혹시 특정 카탈로그를 사용해야지만 생성되는지, 아니면 TBLPROPERTIES 에 설정을 넣어야 하는지 궁금합니다!

아이스버그 V2 부터 도입되었다고 하는데 Docs 를 보아도 해답을 찾기가 어려워서 질문드립니다 !!

**Reply by SML0127** [reply]
안녕하세요, @JeonDaehong 

puffin 파일에 대해 저도 깊게 탐구해보진 않았는데, 질문 주신겸해서 조사 및 테스트 해본 결과 짧게 공유드립니다. 

- puffin 파일의 생성은 자동으로 생성되는게 일반적인것 같습니다. `write.metadata.metrics.default` 설정이 none이 아니며, 통계 정보가 일정 크기 이상을 넘어갈때 별도의 파일로 생성되는 것으로 이해되어요.
- 명시적 생성 또한 가능합니다. 저는 trino에서 `analyze table_name` 쿼리를 수행하였는데 아래 처럼 생성이 가능합니다. spark에도 유사한 기능이 있지 않을까한데 한번 찾아보시는걸 추천드립니다.(참고: [table statistics](https://trino.io/docs/current/connector/iceberg.html#table-statistics))
- trino의 `analyze table_name` 쿼리 결과 생성된 stat 파일은 [apache-datasketches-theta-v1](https://iceberg.apache.org/puffin-spec/#apache-datasketches-theta-v1-blob-type) blob type이고, 칼럼별 몇개의 고유값(`ndv`)이 있는지 대략적으로 파악된 것으로 보입니다.
- 이렇게 생성된 통계 파일은 기본적으로 생성되는 통계 정보보다 더 정교하여, 이후 쿼리 수행시 더 나은 scan planning에 도움이 된다하는데 기회가 되시면 한번 테스트 후 결과 공유주시면 감사하겠습니다.


#### analyze 쿼리 결과 생성된 stat 파일
<img width="662" alt="image" src="https://github.com/user-attachments/assets/5710efb8-f006-4435-aabd-5115b2f41446" />

#### analyze 쿼리 결과 생성된 metadata 파일 내용 
(왼쪽: flink commit으로 생성, 오른쪽: analyze 쿼리로 생성)
<img width="1727" alt="image" src="https://github.com/user-attachments/assets/d6b88fa2-617c-4461-bb57-75cf11c0080b" />

```d
"statistics" : [ {
    "snapshot-id" : 5726817391253084007,
    "statistics-path" : "hdfs://...puffin_test/metadata/20250509_111005_38235_vdzie-421b0217-ef28-4a04-8fa1-413a5675cd46.stats",
    "file-size-in-bytes" : 241981,
    "file-footer-size-in-bytes" : 2004,
    "blob-metadata" : [ {
      "type" : "apache-datasketches-theta-v1",
      "snapshot-id" : 5726817391253084007,
      "sequence-number" : 6,
      "fields" : [ 1 ],
      "properties" : {
        "ndv" : "6373780"
      }
    }, {
      "type" : "apache-datasketches-theta-v1",
      "snapshot-id" : 5726817391253084007,
      "sequence-number" : 6,
      "fields" : [ 2 ],
      "properties" : {
        "ndv" : "42354"
      }
    }, {
      "type" : "apache-datasketches-theta-v1",
      "snapshot-id" : 5726817391253084007,
      "sequence-number" : 6,
      "fields" : [ 3 ],
      "properties" : {
        "ndv" : "2"
      }
    }, {
      "type" : "apache-datasketches-theta-v1",
      "snapshot-id" : 5726817391253084007,
      "sequence-number" : 6,
      "fields" : [ 4 ],
      "properties" : {
        "ndv" : "5223607"
      }
    }, {
      "type" : "apache-datasketches-theta-v1",
      "snapshot-id" : 5726817391253084007,
      "sequence-number" : 6,
      "fields" : [ 5 ],
      "properties" : {
        "ndv" : "5"
      }
    }, {
      "type" : "apache-datasketches-theta-v1",
      "snapshot-id" : 5726817391253084007,
      "sequence-number" : 6,
      "fields" : [ 6 ],
      "properties" : {
        "ndv" : "889166"
      }
    }, {
      "type" : "apache-datasketches-theta-v1",
      "snapshot-id" : 5726817391253084007,
      "sequence-number" : 6,
      "fields" : [ 7 ],
      "properties" : {
        "ndv" : "3391"
      }
    }, {
      "type" : "apache-datasketches-theta-v1",
      "snapshot-id" : 5726817391253084007,
      "sequence-number" : 6,
      "fields" : [ 8 ],
      "properties" : {
        "ndv" : "3147"
      }
    }, {
      "type" : "apache-datasketches-theta-v1",
      "snapshot-id" : 5726817391253084007,
      "sequence-number" : 6,
      "fields" : [ 9 ],
      "properties" : {
        "ndv" : "271668"
      }
    }, {
      "type" : "apache-datasketches-theta-v1",
      "snapshot-id" : 5726817391253084007,
      "sequence-number" : 6,
      "fields" : [ 10 ],
      "properties" : {
        "ndv" : "433828"
      }
    } ]
} ],
```

**Reply by JeonDaehong** [reply]
감사합니다! 퍼핀 파일 테스트 해보고 글 남겨보겠습니다! ㅎㅎㅎ

**Comment by kyungmax** [comment]
iceberg를 공부하는데 정말 큰 도움이 됐습니다. 꽤 긴글인데도 시간가는줄 모르고 읽었네요. 정말 감사합니다 !!
